pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,332 @@
+
+    Returns:
+        list[list[float]]: A nested list of tape results. Each element in
+        the returned list corresponds in order to the provided tapes.
+    """
+    # pylint: disable=unused-argument
+    if max_diff > 1:
+        raise ValueError("The JAX interface only supports first order derivatives.")
+
+    _validate_tapes(tapes)
+
+    for tape in tapes:
+        # set the trainable parameters
+        params = tape.get_parameters(trainable_only=False)
+        tape.trainable_params = qml.math.get_trainable_indices(params)
+
+    parameters = tuple(list(t.get_parameters()) for t in tapes)
+
+    if gradient_fn is None:
+        return _execute_with_fwd(
+            parameters,
+            tapes=tapes,
+            device=device,
+            execute_fn=execute_fn,
+            gradient_kwargs=gradient_kwargs,
+            _n=_n,
+        )
+
+    return _execute(
+        parameters,
+        tapes=tapes,
+        device=device,
+        execute_fn=execute_fn,
+        gradient_fn=gradient_fn,
+        gradient_kwargs=gradient_kwargs,
+        _n=_n,
+    )
+
+
+def _validate_tapes(tapes):
+    """Validates that the input tapes are compatible with JAX support.
+
+    Note: the goal of this validation is to filter out cases where ragged
+    outputs for QNodes may arise. Such QNodes involve creating arrays from
+    ragged nested sequences that can not be handled by JAX.
+
+    Raises:
+        ValueError: if tapes that produce ragged outputs were provided
+    """
+    for t in tapes:
+
+        return_types = [o.return_type for o in t.observables]
+        set_of_return_types = set(return_types)
+        probs_or_sample_measure = Sample in return_types or Probability in return_types
+        if probs_or_sample_measure and len(set_of_return_types) > 1:
+            raise ValueError(
+                "Using the JAX interface, sample and probability measurements cannot be mixed with other measurement types."
+            )
+
+        if Probability in return_types:
+            set_len_wires = set(len(o.wires) for o in t.observables)
+            if len(set_len_wires) > 1:
+                raise ValueError(
+                    "Using the JAX interface, multiple probability measurements need to have the same number of wires specified."
+                )
+
+
+def _execute(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        new_tapes = []
+
+        for t, a in zip(tapes, params):
+            new_tapes.append(t.copy(copy_operations=True))
+            new_tapes[-1].set_parameters(a)
+
+        with qml.tape.Unwrap(*new_tapes):
+            res, _ = execute_fn(new_tapes, **gradient_kwargs)
+
+        if len(res) > 1:
+            res = [jnp.array(r) for r in res]
+        else:
+            res = jnp.array(res)
```

Comment:
is this just to distinguish between a single tape and multiple tapes? Maybe it would be clearer to do `if len(tapes) > 1`?

Reply:
Sure, sounds good!